### PR TITLE
feat(getUser): Get a single user

### PR DESCRIPTION
**Pull request by XROLE**
#### What does this PR do?
- Get single user details from the user's table
#### Description of Task to be completed?
- [x] Retrieve a user from the database by the userId provided
- [x] Send the user to the client with a status code of 200
- [x] Send an error message when there is no such user in the database
- [x] Write test cases to check for successful getting of user
- [x] Write test cases to check when userId is not provided, userId is not a number and userId is not in the database
#### How should this be manually tested?
- Clone the repo, run npm install and then npm start
- Navigate to  GET `http://localhost:5009/api/v1/user/userId` where `userId` is the id of the registered user 
#### Any background context you want to provide?
- N/A
#### What are the relevant pivotal tracker stories?
[163153865]
#### Screenshots (if appropriate)
screen shot 2019-01-14 at 5 38 34 am
#### Questions:
- N/A

### CODE REVIEW

File: controllers/user.js [hunk 1]
```diff
@@ -14,20 +14,10 @@ class Users {
...
-        return res.status(200).json({
-          message: 'User Found',
-          user
-        });
-      })
-      .catch(error => res.status(400).send(error));
+  static async getUser(req, res) {
+    const user = await User.findOne({ where: { id: req.params.userId } });
+    // eslint-disable-next-line no-unused-expressions
+    !user ? res.status(404).json({ message: `No user with userId ${req.params.userId}, Please make sure you have the correct userId.` }) : res.status(200).json({ message: 'success', user });
```
Comment (oreoluwade) [inline controllers/user.js]: The standard we are following currently is to allow only successful operations be evident in the controller.

Also, we are handling errors in the middlewares as I have previously suggested.

Check the` register` controller for clarification.

Comment (XROLE) [reply controllers/user.js]: Ok. I'll implement that now

Comment (XROLE) [reply controllers/user.js]: Feedback implemented

File: controllers/user.js [hunk 2]
```diff
@@ -13,21 +13,11 @@ class Users {
...
-        }
-        return res.status(200).json({
-          message: 'User Found',
-          user
-        });
-      })
-      .catch(error => res.status(400).send(error));
+  static async getUser(req, res) {
+    const user = await User.findOne({ where: { id: req.params.userId } });
+    return res.json({ message: 'success', user });
```
Comment (oreoluwade) [inline controllers/user.js]: I don't think we should include this `message` key-pair value here. Just a user object will be good.

Also, we don't want to send the whole user object back unless absolutely necessary.

What you are sending back will contain `password` for instance, and we know there's no use for an hashed password.

The user object you should send back should be one you have modified to get only necessary details.

Comment (XROLE) [reply controllers/user.js]: OK. Refactoring that now

File: middlewares/isUser.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+     * @param {object} req - The request object
+     * @param {Object} res - The response object
+     * @param {object} next - Method to call another middleware
+     */
+  static async getUser(req, res, next) {
+    const error = new Error('User not found');
+    error.status = 404;
+    const user = await User.findOne({ where: { id: req.params.userId } });
+    // eslint-disable-next-line no-unused-expressions
+    !user ? next(error) : next();
```
Comment (oreoluwade) [inline middlewares/isUser.js]: Endeavor to always explicitly return something from your functions in express. You may come across the `cannot set headers after...` error if you don't

File: middlewares/isUser.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+/* eslint-disable no-trailing-spaces */
```
Comment (oreoluwade) [inline middlewares/isUser.js]: Why did you disable this rule?

File: middlewares/isUser.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+*/
+export default class {
+  /**
+     * Check if a user exist in the database
+     * @constructor
+     * @param {object} req - The request object
+     * @param {Object} res - The response object
+     * @param {object} next - Method to call another middleware
+     */
+  static async getUser(req, res, next) {
```
Comment (oreoluwade) [inline middlewares/isUser.js]: Is there a specific reason you abstracted this middleware to its own file as opposed to including it in the existing `userValidations` module?

File: test/user.test.js [hunk 6]
```diff
@@ -230,4 +230,42 @@ describe('USER TEST SUITE', () => {
...
+    it('Should get a user with valid user id present in database', (done) => {
+      chai.request(server)
+        .get('/api/v1/user/1')
+        .end((err, res) => {
+          expect(res.status).to.equal(200);
+          expect(res.body.message).to.equal('success');
+          done();
+        });
+    });
+    it('Should should not get a user when the id is not found in the database',
```
Comment (oreoluwade) [inline test/user.test.js]: Please, proofread the description for each of your specs.

File: middlewares/isUser.js (hunk 4 above)
Comment (XROLE) [reply middlewares/isUser.js]: I did that while debugging and I forgot to enable it back. But I've taken care of it now

File: middlewares/isUser.js (hunk 5 above)
Comment (XROLE) [reply middlewares/isUser.js]: not really, just that I taught it was meant to validate user registrations alone. But I'l merge them together right away

File: test/user.test.js (hunk 6 above)
Comment (XROLE) [reply test/user.test.js]: Ok. Will do that right away

File: middlewares/isUser.js (hunk 3 above)
Comment (XROLE) [reply middlewares/isUser.js]: Owh Thanks.

File: test/user.test.js (hunk 6 above)
Comment (XROLE) [reply test/user.test.js]: Thanks for that. I noticed and rectified the double appearance of `should` in about three test cases.

File: middlewares/isUser.js (hunk 5 above)
Comment (XROLE) [reply middlewares/isUser.js]: Done

File: middlewares/userValidations.js [hunk 7]
```diff
@@ -2,6 +2,7 @@ import Sequelize from 'sequelize';
 import validation from 'validator';
 import models from '../models';
 
+
```
Comment (oreoluwade) [inline middlewares/userValidations.js]: Remove the extra line here

File: middlewares/userValidations.js [hunk 8]
```diff
@@ -10,6 +11,14 @@ const nonEmptyParams = ['username', 'email', 'password'];
 
 
 export default {
+  async isUser(req, res, next) {
+    const error = new Error('User not found');
+    error.status = 404;
+    const user = await User.findOne({ where: { id: req.params.userId } });
+    // eslint-disable-next-line no-unused-expressions
```
Comment (oreoluwade) [inline middlewares/userValidations.js]: Remove this eslint disable, then return the expression on the next line

File: middlewares/userValidations.js [hunk 9]
```diff
@@ -112,3 +121,6 @@ export default {
   }
 
 };
+
```
[the remark this answers is not in the document]
Comment (oreoluwade) [reply middlewares/userValidations.js]: Please attend to this.

File: middlewares/userValidations.js [hunk 10]
```diff
@@ -112,3 +121,6 @@ export default {
...
+
+
```
[the remark this answers is not in the document]
Comment (oreoluwade) [reply middlewares/userValidations.js]: This too.

**Review by oreoluwade — CHANGES_REQUESTED** [review]

File: test/user.test.js [hunk 11]
```diff
@@ -230,4 +230,41 @@ describe('USER TEST SUITE', () => {
...
+          .get('/api/v1/user/')
+          .end((err, res) => {
+            expect(res.status).to.equal(404);
+            done();
+          });
+      });
+    it('Should not get a user when the id is not a number',
+      (done) => {
+        chai.request(server)
+          .get('/api/v1/user/')
```
Comment (oreoluwade) [inline test/user.test.js]: Change the route here to `'/api/v1/user/my-user'`.

**Review by oreoluwade — APPROVED** [review]

**Review by oreoluwade — APPROVED** [review]

**Pull request merged**